cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,312 @@
+		}
+		if err := installPlugin(ctx, tmpFile.Name(), opts.inputChecksum, opts.force); err != nil {
+			return fmt.Errorf("plugin installation failed: %w", err)
+		}
+		return nil
+	default:
+		return errors.New("plugin installation failed: unknown plugin source type")
+	}
+}
+
+// installPlugin installs the plugin given plugin source path
+func installPlugin(ctx context.Context, inputPath string, inputChecksum string, force bool) error {
+	// sanity check
+	inputFileStat, err := os.Stat(inputPath)
+	if err != nil {
+		return err
+	}
+	if !inputFileStat.Mode().IsRegular() {
+		return fmt.Errorf("%s is not a valid file", inputPath)
+	}
+	// checksum check
+	if inputChecksum != "" {
+		if err := osutil.ValidateSHA256Sum(inputPath, inputChecksum); err != nil {
+			return err
+		}
+	}
+	// install the plugin based on file type
+	fileType, err := osutil.DetectFileType(inputPath)
+	if err != nil {
+		return err
+	}
+	switch fileType {
+	case notationplugin.MediaTypeZip:
+		rc, err := zip.OpenReader(inputPath)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		return installPluginFromFS(ctx, rc, force)
+	case notationplugin.MediaTypeGzip:
+		// when file is gzip, required to be tar
+		return installPluginFromTarGz(ctx, inputPath, force)
+	default:
+		// input file is not in zip or gzip, try install directly
+		installOpts := plugin.CLIInstallOptions{
+			PluginPath: inputPath,
+			Overwrite:  force,
+		}
+		return installPluginWithOptions(ctx, installOpts)
+	}
+}
+
+// installPluginFromFS extracts, validates and installs the plugin files
+// from a fs.FS
+//
+// Note: zip.ReadCloser implments fs.FS
+func installPluginFromFS(ctx context.Context, pluginFs fs.FS, force bool) error {
+	// set up logger
+	logger := log.GetLogger(ctx)
+	root := "."
+	// extracting all regular files from root into tmpDir
+	tmpDir, err := os.MkdirTemp("", notationPluginTmpDir)
+	if err != nil {
+		return fmt.Errorf("failed to create notationPluginTmpDir: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	if err := fs.WalkDir(pluginFs, root, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		fName := d.Name()
+		if d.IsDir() && fName != root { // skip any dir in the fs except root
+			return fs.SkipDir
+		}
+		info, err := d.Info()
+		if err != nil {
+			return err
+		}
+		// only accept regular files.
+		// it is required by github-advanced-security to check for `..` in fName
+		if !info.Mode().IsRegular() || strings.Contains(fName, "..") {
+			return nil
+		}
+		logger.Debugf("Extracting file %s...", fName)
+		rc, err := pluginFs.Open(path)
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		tmpFilePath := filepath.Join(tmpDir, fName)
+		return osutil.CopyFromReaderToDir(rc, tmpFilePath, info.Mode())
+	}); err != nil {
+		return err
+	}
+	// install core process
+	installOpts := plugin.CLIInstallOptions{
+		PluginPath: tmpDir,
+		Overwrite:  force,
+	}
+	return installPluginWithOptions(ctx, installOpts)
+}
```

Comment:
This part can be refactored into `notation-go` by having a new field like `FS fs.FS` in `CLIInstallOptions`.

Reply:
@shizhMSFT How about we move it to notation-go in the next iteration? (It won't be a breaking change.)

Reply:
I think you can start a new PR in `notation-go`. If that PR merges first, we can update that in this PR. Otherwise, we can do it after merging this PR.

Reply:
@shizhMSFT Could you approve this PR first as the 1st iteration of the feature? Making the above changes is not a quick PR, it's actually the next iteration of the command. The update is in progress, created this issue to track: https://github.com/notaryproject/notation/issues/852